compiler/src/dotty/tools/dotc/core/GadtConstraint.scala

```diff
@@ -63,31 +120,281 @@ final class ProperGadtConstraint private(
   private var myConstraint: Constraint,
   private var mapping: SimpleIdentityMap[Symbol, TypeVar],
   private var reverseMapping: SimpleIdentityMap[TypeParamRef, Symbol],
-  private var wasConstrained: Boolean
+  private var pathDepMapping: SimpleIdentityMap[PathType, SimpleIdentityMap[Symbol, TypeVar]],
+  private var pathDepReverseMapping: SimpleIdentityMap[TypeParamRef, TypeRef],
+  private var wasConstrained: Boolean,
+  private var myScrutineePath: TermRef | Null,
+  private var myUnionFind: SimpleIdentityMap[PathType, PathType],
+  private var myPatternSkolem: SkolemType | Null,
 ) extends GadtConstraint with ConstraintHandling {
   import dotty.tools.dotc.config.Printers.{gadts, gadtsConstr}
 
   def this() = this(
     myConstraint = new OrderingConstraint(SimpleIdentityMap.empty, SimpleIdentityMap.empty, SimpleIdentityMap.empty),
     mapping = SimpleIdentityMap.empty,
     reverseMapping = SimpleIdentityMap.empty,
-    wasConstrained = false
+    pathDepMapping = SimpleIdentityMap.empty,
+    pathDepReverseMapping = SimpleIdentityMap.empty,
+    wasConstrained = false,
+    myScrutineePath = null,
+    myUnionFind = SimpleIdentityMap.empty,
+    myPatternSkolem = null,
```

Comment:
Why is it necessary to keep this reference here?

Reply:
Yes we have to keep track of the skolem created as a placeholder for the pattern path, so that when the real pattern path is available we can substitute the placeholder with the actual path. But maybe we should move this reference to somewhere else?